File: from-1501-to-1600/1593-split-a-string-into-the-max-number-of-unique-substrings/solution.go
func maxUniqueSplit(s string) int {
    var dfs func(idx int, currSet map[string]bool) int
    dfs = func(idx int, currSet map[string]bool) int {
        if idx == len(s) {
            return 0
        }
        result := 0
        for j := idx; j < len(s); j ++ {
            substr := s[idx: j + 1]
            if currSet[substr] {
                continue
            }
            currSet[substr] = true
            result = max(result, dfs(j + 1, currSet) + 1)
            currSet[substr] = false
        }
        return result
    }
    return dfs(0, make(map[string]bool))
}

File: from-1501-to-1600/1593-split-a-string-into-the-max-number-of-unique-substrings/solution.py
class Solution:
    def maxUniqueSplit(self, s: str) -> int:
        def dfs(i, currSet):
            if i == len(s):
                return 0
            result = 0
            for j in range(i, len(s)):
                substr = s[i:j + 1]
                if substr in currSet:
                    continue
                currSet.add(substr)
                result = max(result, dfs(j + 1, currSet) + 1)
                currSet.remove(substr)

            return result
        return dfs(0, set())